go/enclave/nodetype/sequencer.go

```diff
@@ -168,16 +183,25 @@ func (s *sequencer) createNewHeadBatch(l1HeadBlock *common.L1Block, skipBatchIfE
 		return fmt.Errorf("attempted to create batch on top of batch=%s. With l1 head=%s", headBatch.Hash(), l1HeadBlock.Hash())
 	}
 
-	stateDB, err := s.storage.CreateStateDB(headBatch.Hash())
-	if err != nil {
-		return fmt.Errorf("unable to create stateDB for selecting transactions. Batch: %s Cause: %w", headBatch.Hash(), err)
-	}
-
 	// todo (@stefan) - limit on receipts too
 	limiter := limiters.NewBatchSizeLimiter(s.settings.MaxBatchSize)
-	transactions, err := s.mempool.CurrentTxs(stateDB, limiter)
-	if err != nil {
-		return err
+	pendingTransactions := s.mempool.PendingTransactions()
```

Comment:
I thought the same thing, but I didn't want to pass the limiter to the tx pool. Do you think the limiter should be in the txpool ? So that we can lazily build the batch ?

Reply:
Very similar..! Cycle through each lazy tx, resolve it, and then add it to the block until there's it hits the block gas.

```
for {
		// Check interruption signal and abort building if it's fired.
		if interrupt != nil {
			if signal := interrupt.Load(); signal != commitInterruptNone {
				return signalToErr(signal)
			}
		}
		// If we don't have enough gas for any further transactions then we're done.
		if env.gasPool.Gas() < params.TxGas {
			log.Trace("Not enough gas for further transactions", "have", env.gasPool, "want", params.TxGas)
			break
		}
		// Retrieve the next transaction and abort if all done.
		ltx := txs.Peek()
		if ltx == nil {
			break
		}
		tx := ltx.Resolve()
		if tx == nil {
			log.Warn("Ignoring evicted transaction")

			txs.Pop()
			continue
		}
		...
		case errors.Is(err, nil):
			// Everything ok, collect the logs and shift in the next transaction from the same account
			coalescedLogs = append(coalescedLogs, logs...)
			env.tcount++
			txs.Shift()
		```